debugging/debugging_models/debugging_components/debugging_model/test_modelshell.py

```diff
@@ -0,0 +1,419 @@
+'''
+The model_shell class object's forward function uses 3 different forward functions from the 3 different class objects that are passed to it.
+These functions perform overarching tasks that are 1) embedder, 2) core model, 3) model head.
+
+Generally...
+...the embedder is a class object that generally takes in a tensor (B, T) and returns a tensor (B, T, H).
+...the core model is a class object that takes in a tensor (B, T, H) and returns a tensor (B, T, H).
+...the model head is a class object that takes in a tensor (B, T, H) and returns a tensor (B, T, V).
+
+
+This pytest will ensure that these outputs are correctly shaped and are not nan. 
+'''
+
+import pytest
+import torch
+
+from models.core_models import GenericFFNSharedTransfomer, GenericTransformer
+from models.embedding_models import GenericEmbedder
+from models.experimental.byte_level.embedding_model import ByteLevelEmbedder
+from models.experimental.byte_level.model_heads import ByteLevelDecoder
+from models.experimental.hugging_face import HFEmbedder, HFLMHead, HFTransformerCore
+from models.experimental.next_thought.embedding_models import HierarchicalEncoder
+from models.experimental.next_thought.model_heads import VariableLengthLatentDecoder
+from models.experimental.next_thought.core_models import BaselineCoreModel, Conv1dCoreModel
+from models.model_heads import AutoregressiveLMHead
+
+
+
+################################################### embedders ###################################################
+
+def test_genericembedder():
+    """
+    Test the embedding model.
+    """
+    model_cfg = {
+        "embedder": {
+            "tokenizer_type": "gpt2",
+            "embedding_model_type": "generic",
+            "dataset_name": "stlm"
+        },
+        "hidden_dim": 64,
+        "context_window": 64,
+        "vocab_size": 50257,
+        "positional_encoding_type": "rope",
+        'batch_size': 1
+    }
+
+    input_ids = torch.randint(0, model_cfg['vocab_size'], (model_cfg['batch_size'], model_cfg['context_window']))
+
+    ## build the embedder
+    embedder = GenericEmbedder(model_cfg)
+
+    ## get the output
+    res = embedder(input_ids)
+
+    ## 1. ensure the output is float32
+    assert res.dtype == torch.float32
+    ## 2. ensure the output is not nan
+    assert not torch.isnan(res).all()
+    ## 3. ensure the output shape is correct
+    assert res.shape == (model_cfg['batch_size'], model_cfg['context_window'], embedder.token_embedder.embedding_dim)
```

Comment:
these aren't really testing the modelshell, they are testing the individual components...

Reply:
which is fine but should be in seperate test scripts. i.e. should have test_embedder, test_core_model, test_lm_head as files

Reply:
and these are just testing the forward pass. that is fine for the core model and lm head, but the embedder interface has quite a few methods (padding, truncating, inference vs forward, etc. that should be checked